src/language/templates/ast/all.hpp

```diff
@@ -20,6 +20,9 @@
  */
 
 #include "ast/ast.hpp"
+{% for header in node_info.CIRCULAR_DEPENDECIES %}
```

Comment:
Without this this is the issue I get:
```
In file included from /Users/magkanar/nmodl_local/src/parser/nmodl_driver.cpp:11:
In file included from /Users/magkanar/nmodl_local/src/lexer/nmodl_lexer.hpp:16:
In file included from /Users/magkanar/nmodl_local/src/parser/nmodl.yy:31:
/Users/magkanar/nmodl_local/build/src/ast/all.hpp:7233:10: error: member access into incomplete type 'nmodl::ast::AssignedDefinition'
        n->set_parent(this);
         ^
```
I believe that the issue is not with forward declaring the class, since this is done in `ast_decl.hpp` but the method `AssignedDefinition::set_parent()`. To solve this issue we should forward declare all the methods then, which I am not convinced that is a better option

Reply:
Since @tristan0x was working on these things, I really would like to hear his opinion too?

Reply:
Moving the definitions of these member methods to `ast.cpp` fixes most of the compilation issues but for the template member function 

```c++
/**
 * \brief Insert members to statements
 */
template <class NodeType, class InputIterator>
void StatementBlock::insert_statement(StatementVector::const_iterator position,
                                      NodeType& to,
                                      InputIterator first,
                                      InputIterator last) {
    for (auto it = first; it != last; ++it) {
        auto& n = *it;
        // set parents
        n->set_parent(this);
    }
    auto pos_it = const_iter_cast(statements, position);
    auto first_it = const_iter_cast(to, first);
    auto last_it = const_iter_cast(to, last);
    statements.insert(pos_it, first_it, last_it);
}
```

Possible solutions:
1. put the definition of this member method after the definitions of all the Ast classes in `ast/all.hpp`.
1. Put its definition in `ast/ast.cpp` and use explicit template instantiations for all possible cases
1. Question the usefulness of this member function:
    1. `const_iter_cast` is a workaround for an issue that we might not have anymore (see definition of this helper function)
    1. `first` and `last` are always iterators over a subclass of statements. There should be an easy way to express the same thing without using a template.

Reply:
Thanks a lot for your input @tristan0x !
1. Unfortunately this is not possible as fas as I know because this would mean moving all the `Statement`s on top of `Expression`s, which would fail again due to other circular dependencies
2. The definition of this for every case in `ast/ast.cpp` is something I would like to avoid, since I don't believe it's very nice
3. This might be a proper solution and I need to investigate it.
Thanks a lot again!

Reply:
fixed in #351